### PR TITLE
Call Token

### PR DESCRIPTION
**Pull request by erikzhang**

### CODE REVIEW

**Comment by shargon** [comment]
How It will be used?

File: src/neo-vm/EvaluationStack.cs [hunk 1]
```diff
@@ -57,6 +57,8 @@ internal void Insert(int index, StackItem item)
 
         internal void MoveTo(EvaluationStack stack, int count = -1)
         {
+            if (count < -1 || count > innerList.Count)
+                throw new ArgumentOutOfRangeException(nameof(count));
```
Comment (erikzhang) [inline src/neo-vm/EvaluationStack.cs:61]: This check will be run in `CopyTo`.

**Comment by erikzhang** [comment]
> How It will be used?

I will add `MethodToken` table in NefFile.

```
0x0001: <hash> <methodName> <paramCount> <rvCount> <flags>
0x0002: <hash> <methodName> <paramCount> <rvCount> <flags>
0x0003: <hash> <methodName> <paramCount> <rvCount> <flags>
...
```

Then we can call other contracts by `CALLT`:

```
PUSH <param1>
PUSH <param2>
CALLT 0x0002
```

**Comment by shargon** [comment]
> I will add MethodToken table in NefFile.

Then maybe we will need more than 4 bytes in NefFile.

**Comment by erikzhang** [comment]
Why? It only takes up 1 byte.

**Comment by shargon** [comment]
> Why? It only takes up 1 byte.

3 more seems weird :P

File: src/neo-vm/ExecutionEngine.cs [hunk 2]
```diff
@@ -299,6 +299,13 @@ private void ExecuteInstruction()
                         ExecuteCall(x.Position);
                         break;
                     }
+                case OpCode.CALLT:
+                    {
+                        ExecutionContext context_current = CurrentContext;
+                        IToken token = LoadToken(instruction.TokenU16, out ExecutionContext context_loaded);
+                        context_current.EvaluationStack.MoveTo(context_loaded.EvaluationStack, token.ParametersCount);
```
Comment (shargon) [inline src/neo-vm/ExecutionEngine.cs:306]: We should copy `ReturnValuesCount` when unload the context?

Comment (erikzhang) [reply src/neo-vm/ExecutionEngine.cs]: Done.

File: src/neo-vm/ExecutionContext.cs [hunk 3]
```diff
@@ -11,6 +11,16 @@ public sealed partial class ExecutionContext
...
 
+        /// <summary>
+        /// Number of parameters
+        /// </summary>
+        public ushort ParametersCount { get; }
+
+        /// <summary>
+        /// Number of items to be returned
+        /// </summary>
+        public int RVCount { get; }
```
Comment (shargon) [inline src/neo-vm/ExecutionContext.cs:22]: short?

File: src/neo-vm/ExecutionEngine.cs [hunk 4]
```diff
@@ -361,7 +368,15 @@ private void ExecuteInstruction()
                         ExecutionContext context_pop = InvocationStack.Pop();
                         EvaluationStack stack_eval = InvocationStack.Count == 0 ? ResultStack : InvocationStack.Peek().EvaluationStack;
                         if (context_pop.EvaluationStack != stack_eval)
-                            context_pop.EvaluationStack.CopyTo(stack_eval);
+                        {
+                            int rvcount = context_pop.RVCount;
+                            if (rvcount < 0)
+                                rvcount = context_pop.EvaluationStack.Count;
+                            else if (context_pop.EvaluationStack.Count != rvcount)
```
Comment (shargon) [inline src/neo-vm/ExecutionEngine.cs:375]: These checks will also be done in `EvaluationStack.CopyTo`

File: src/neo-vm/ExecutionEngine.cs [hunk 5]
```diff
@@ -1371,21 +1386,26 @@ protected virtual void LoadContext(ExecutionContext context)
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        protected ExecutionContext CreateContext(Script script, int initialPosition = 0)
+        protected ExecutionContext CreateContext(Script script, ushort pcount, int rvcount, int initialPosition)
```
Comment (shargon) [inline src/neo-vm/ExecutionEngine.cs:1389]: Default value was removed, intencional?

File: src/neo-vm/ExecutionContext.cs (hunk 3 above)
Comment (erikzhang) [reply src/neo-vm/ExecutionContext.cs:22]: https://github.com/neo-project/neo-vm/blob/6fc25a0239c411c815c13994978c2a0fb608dfc0/src/neo-vm/ExecutionContext.cs#L94-L97

File: src/neo-vm/ExecutionEngine.cs (hunk 4 above)
Comment (erikzhang) [reply src/neo-vm/ExecutionEngine.cs]: Done.

File: src/neo-vm/ExecutionEngine.cs (hunk 5 above)
Comment (erikzhang) [reply src/neo-vm/ExecutionEngine.cs:1385]: Yes.

**Review by shargon — APPROVED** [review]

**Pull request merged**